store_init.go

```diff
@@ -0,0 +1,278 @@
+		return nil, fmt.Errorf(
+			"code was tested on PostgreSQL version %d only: change the source if you want to try running on the older version %d",
+			minimumPgVersion, currentPgVersion,
+		)
+	}
+
+	dbSettingsExpected := "Encoding:SQL_ASCII Collate:C Ctype:C"
+	var dbSettings string
+	if err := dbPool.QueryRow(
+		ctx,
+		`
+		SELECT 'Encoding:' || pg_encoding_to_char(encoding) || ' Collate:' || datcollate || ' Ctype:' || datctype
+			FROM pg_database
+		WHERE datname = current_database()
+		`,
+	).Scan(&dbSettings); err != nil {
+		return nil, err
+	}
+	if dbSettings != dbSettingsExpected {
+		return nil, fmt.Errorf(
+			"unexpected database settings: you must create your database with something like `%s` for reliable and performant binary storage\n Current settings: %s\nExpected settings: %s",
+			`CREATE DATABASE {{name}} ENCODING='SQL_ASCII' LC_COLLATE='C' LC_CTYPE='C' TEMPLATE='template0'`,
+			dbSettings, dbSettingsExpected,
+		)
+	}
+
+	// check whether we are talking to a slave, and force RO to shut off tipset-tracking automatically
+	if dbbs.isWritable {
+		if _, err := dbPool.Exec(ctx, fmt.Sprintf(
+			"CREATE TEMPORARY TABLE %s ( pk INTEGER ) ON COMMIT DROP",
+			"tmptable_"+randBytesAsHex(),
+		)); err != nil {
+			dbbs.isWritable = false
+		}
+	}
+
+	dbbs.lru, err = cidkeyedlru.NewCidKeyedLruCache(dbbs.lruSizeBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to instantiate cache: %w", err)
+	}
+
+	if cfg.LogDetailedAccess {
+		if dbbs.instanceNamespace == "" {
+			return nil, xerrors.New("detailed access logging not possible without an instance-specific chainsync namespace")
+		}
+		if !dbbs.isWritable {
+			return nil, xerrors.New("detailed access logging not possible on a read-only store")
+		}
+		dbbs.accessLogsDetailed = make(map[accessUnit]int64, 65535)
+	}
+
+	neededDDL, err := dbbs.neededDeployDDL(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to determine current deployed schema status: %w", err)
+	}
+
+	if neededDDL != nil {
+
+		if !dbbs.isWritable {
+			if dbbs.instanceNamespace != "" {
+				return nil, fmt.Errorf(
+					"unable to continue: connecting in ReadOnly mode requires that the view %s.current_head and all associates dependencies have been already deployed",
+					dbbs.instanceNamespace,
+				)
+			}
+			return nil, xerrors.New(
+				"unable to continue: connecting in ReadOnly mode requires that the fil_common_base schema has been already deployed",
+			)
+		}
+
+		if !cfg.AutoUpdateSchema {
+			return nil, xerrors.New(
+				"unable to continue: currently-deployed schema does not (partially) match the expected state, and AutoUpdateSchema is not enabled",
+			)
+		}
+
+		if err = dbbs.deploy(ctx, neededDDL); err != nil {
+			return nil, err
+		}
+	}
+
+	// not RO and everything is deployed: let's make sure we can:
+	// - EX-lock the namespace or quit
+	// - then auto-SH-lock it going forward
+	if dbbs.isWritable && dbbs.instanceNamespace != "" {
```

Comment:
I don't understand this handover sequence. You attempt to take an exclusive lock then wait on a shared lock in a new connection until the exclusive lock connection has dropped, then retry the exclusive lock. What is the purpose of the second try?

Reply:
@iand the thinking is that this snippet:
https://github.com/filecoin-project/go-bs-postgres-chainnotated/blob/8b1f2e9296c4346ca7647889cde5d2ecae123f8e/store_init.go#L211-L220
could get tripped up by a *different* piece of code ( perhaps not even written by me ) grabbing a shared lock as soon as the other exlock quits.
Basically I am making sure that the first connection from the new pool can properly upgrade itself, indicating that yes indeed the entire dance worked. Perhaps I am being overly paranoid and should just remove the entire "double-check" section.

Thoughts?

Reply:
@iand I am a muppet, after sleeping on it - you are 100% correct
Fixed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/0edaa329d98